rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -523,16 +523,35 @@ std::shared_ptr<rclcpp::GenericSubscription>
 RecorderImpl::create_subscription(
   const std::string & topic_name, const std::string & topic_type, const rclcpp::QoS & qos)
 {
-  auto subscription = node->create_generic_subscription(
+  std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>,
+    const rclcpp::MessageInfo &)> recvFun;
+
+  if (record_options_.use_sim_time) {
+    return node->create_generic_subscription(
+      topic_name,
+      topic_type,
+      qos,
+      [this, topic_name, topic_type](std::shared_ptr<const rclcpp::SerializedMessage> message) {
+        if (!paused_.load()) {
+          writer_->write(
+            message, topic_name, topic_type, node->now());
+        }
+      });
+  }
+
+  return node->create_generic_subscription(
     topic_name,
     topic_type,
     qos,
-    [this, topic_name, topic_type](std::shared_ptr<const rclcpp::SerializedMessage> message) {
+    [this, topic_name, topic_type](std::shared_ptr<const rclcpp::SerializedMessage> message,
+    const rclcpp::MessageInfo & mi) {
       if (!paused_.load()) {
-        writer_->write(message, topic_name, topic_type, node->get_clock()->now());
+        writer_->write(
+          message, topic_name, topic_type,
+          rclcpp::Time(mi.get_rmw_message_info().received_timestamp),
+          rclcpp::Time(mi.get_rmw_message_info().source_timestamp));
       }
     });
-  return subscription;
 }
```

Comment:
@jmachowinski I think we can simplify the implementation of this function and reduce delta by doing it like this
```cpp
std::shared_ptr<rclcpp::GenericSubscription>
RecorderImpl::create_subscription(
  const std::string & topic_name, const std::string & topic_type, const rclcpp::QoS & qos)
{
  return node->create_generic_subscription(
    topic_name,
    topic_type,
    qos,
    [this, topic_name, topic_type](std::shared_ptr<const rclcpp::SerializedMessage> message,
    const rclcpp::MessageInfo & mi) {
      if (!paused_.load()) {
        writer_->write(
          message, topic_name, topic_type,
          record_options_.use_sim_time ?
          node->get_clock()->now().nanoseconds() : mi.get_rmw_message_info().received_timestamp),
          mi.get_rmw_message_info().source_timestamp);
      }
    });
}
```

Reply:
This would add an additional if on every message receive.

Reply:
Hm, do we want to set the rmw send timestamp in the sim case ? It will be in a different time domain

Reply:
> I also noticed that you changed `node->get_clock()->now()` to the `node->now()` as far as I remember they are not the same and we had some issue before with it and deliberately using `node->get_clock()->now()` for simulation time.

There should be literally not difference
```c++
rclcpp::Time
Node::now() const
{
  return node_clock_->get_clock()->now();
}
``` 
```c++
rclcpp::Clock::ConstSharedPtr
Node::get_clock() const
{
  return node_clock_->get_clock();
}
```

Reply:
@jmachowinski Thanks for elaborating and digging deeper into the rclcpp node's clock routines.
I agree there are no differences. It seems I messed up with something else.